lambda/matching.ml

```diff
@@ -1178,166 +1194,76 @@ let rec split_or argo cls args def =
-            let yes = List.rev rev_yes and no = List.rev rev_no in
-            match no with
-            | [] ->
-                ( { me = Pm { cases = yes; args; default = def };
-                    matrix = as_matrix yes;
-                    top_default = def
-                  },
-                  k )
-            | cl :: rem -> (
-                match yes with
-                | [] ->
-                    (* Could not succeeed in raising up a constr matching up *)
-                    split_noex [ cl ] [] rem
-                | _ ->
-                    let { me = next; matrix; top_default = def }, nexts =
-                      split_noex [ cl ] [] rem
-                    in
-                    let idef = next_raise_count () in
-                    let def = Default_environment.cons matrix idef def in
-                    ( { me = Pm { cases = yes; args; default = def };
-                        matrix = as_matrix yes;
-                        top_default = def
-                      },
-                      (idef, next) :: nexts )
-              )
-          )
-      and split_noex rev_yes rev_no = function
-        | ([], _) :: _ -> assert false
-        | [ ((ps, _) as cl) ] when List.for_all group_var ps && rev_yes <> []
-          ->
-            (* This enables an extra division in some frequent cases:
-               last row is made of variables only *)
-            split_noex rev_yes (cl :: rev_no) []
-        | ((p :: _, _) as cl) :: rem ->
-            if (not (group p)) && safe_before cl rev_no then
-              split_noex (cl :: rev_yes) rev_no rem
-            else
-              split_noex rev_yes (cl :: rev_no) rem
-        | [] -> (
-            let yes = List.rev rev_yes and no = List.rev rev_no in
-            match no with
-            | [] -> precompile_var args yes def k
-            | cl :: rem ->
-                let { me = next; matrix; top_default = def }, nexts =
-                  split_ex [ cl ] [] rem
-                in
-                let idef = next_raise_count () in
-                precompile_var args yes
-                  (Default_environment.cons matrix idef def)
-                  ((idef, next) :: nexts)
-          )
-      in
-      match cls with
-      | ((p :: _, _) as cl) :: rem ->
-          if group p then
-            split_ex [ cl ] [] rem
-          else
-            split_noex [ cl ] [] rem
-      | _ -> assert false
-    )
+        do_not_precompile
+    in
+    match no with
+    | [] -> precompile_group args yes def k
+    | _ ->
+        let { me = next; matrix; top_default = def }, nexts = split no in
+        let idef = next_raise_count () in
+        precompile_group args yes
+          (Default_environment.cons matrix idef def)
+          ((idef, next) :: nexts)
+  and should_split group_discr =
+    match group_discr.pat_desc with
+    | Tpat_construct (_, { cstr_tag = Cstr_extension _ }, _) ->
+        (* it is unlikely that we will raise anything, so we split now *)
```

Comment:
If I am not mistaken, the unlikely here requires the sensible assumption that the distributions of constructor head pattern in the pm is not dominated by a single constructor. In other words, we discard the case
```
A _ _ ... 
_ _ _ ....
A _ _ ...
_ _ _ ....
A _ _ ...
...
```
as improbable. Am i wrong?

Reply:
It's actually even more specific than that. Not only would you need to have the same constructor at the head of later rows, but you would also need to be able to lift these rows up!
That is, they would have to be incompatible with all the other rows that are in between: which won't happen if there is another constructor with the same arity in the middle.

We deem that very improbable indeed.